File: lib/engine/justwatch.dart
// JustWatch 适配器（逆向其网页端 GraphQL，纯 Dart 移植）
//
// 接口要点（2026-09 逆向验证）：
// - 未使用的变量会返回 422，查询里每个变量都必须出现；
// - translation(language: "en") 是字符串标量，字面量必须带引号小写；
// - language=zh 时 content.title 直接是中文译名（无译名回退原名）；
// - searchQuery 多词会稀释结果，只传主词；类型用 genres 短码（hrr=恐怖…）；
// - releaseYear 是 IntFilter，字段名 min/max；offers 的 platform 枚举用 IOS；
// - posterUrl 相对路径，{profile} 占位符替换为 s332 加 images 前缀。
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'http.dart';
import 'models.dart';
import 't2s.dart';
import 'taxonomy.dart';

const _gqlUrl = 'https://apis.justwatch.com/graphql';
const _imgPrefix = 'https://images.justwatch.com';

const _headers = {
  'Content-Type': 'application/json',
  'Origin': 'https://www.justwatch.com',
  'Referer': 'https://www.justwatch.com/',
  'User-Agent': kBrowserUa,
};

const _query = r'''
query($f: TitleFilter, $c: Country!, $l: Language!, $p: Platform!, $n: Int!) {
  popularTitles(filter: $f, country: $c, language: $l, first: $n) {
    edges { node {
      id
      content(country: $c, language: $l) {
        title originalTitle fullPath shortDescription
        originalReleaseYear
        scoring { imdbScore imdbVotes }
        externalIds { imdbId tmdbId }
        genres { shortName translation(language: "en") }
        posterUrl(format: JPG)
      }
      offers(country: $c, platform: $p) {
        monetizationType standardWebURL presentationType
        package { shortName technicalName clearName }
        retailPrice(language: $l) currency
      }
    } }
  }
}''';

const _monetizationKind = {
  'FREE': 'free', 'ADS': 'free', // 免费广告支撑平台——没会员也能看
  'FLATRATE': 'sub', // 订阅制
  'RENT': 'rent', 'BUY': 'buy',
  'CINEMA': 'other', 'DOWNLOAD': 'buy', 'DVD': 'other',
};

const jwGenreCodes = {
  'horror': 'hrr', 'thriller': 'trl', 'sci-fi': 'scf', 'fantasy': 'fnt',
  'comedy': 'cmy', 'action': 'act', 'animation': 'ani', 'adventure': 'act',
  'mystery': 'trl', 'disaster': '',
};

class JustWatchAdapter {
  final status = SourceStatus('justwatch', 'scrape');

  Future<List<dynamic>> _fetch(Map<String, dynamic> titleFilter, String country,
      {int n = 24, bool record = true}) async {
    final variables = {'f': titleFilter, 'c': country, 'l': 'zh', 'p': 'IOS', 'n': n};
    final sw = Stopwatch()..start();
    try {
      final r = await throttleHost('apis.justwatch.com').then((_) => httpPost(
            Uri.parse(_gqlUrl),
            headers: _headers,
            body: jsonEncode({'query': _query, 'variables': variables}),
          ));
      if (r.statusCode != 200) throw Exception('HTTP ${r.statusCode}');
      final d = jsonDecode(utf8.decode(r.bodyBytes)) as Map<String, dynamic>;
      final data = (d['data'] as Map?)?['popularTitles'];
      if (d['errors'] != null && data == null) {
        final errs = d['errors'] as List;
        throw Exception((errs.first as Map)['message']);
      }
      final edges = ((data as Map?)?['edges'] as List?) ?? [];
      if (record) status.recordOk(sw.elapsedMilliseconds * 1.0);
      return edges.whereType<Map>().map((e) => e['node']).whereType<Map>().toList();
    } catch (e) {
      if (record) status.recordFail(e.toString());
      rethrow;
    }
  }

  Future<List<Movie>> search(ParsedQuery pq, String country) async {
    // 「所有国家」：美区发现 + 英/日区补查观看渠道
    final multiRegion = country == 'ALL';
    final baseCountry = multiRegion ? 'US' : country;

    final terms = <String>[];
    for (final lang in ['en', 'ja', 'ko']) {
      final l = pq.termsByLang[lang] ?? const [];
      if (l.isNotEmpty && !terms.contains(l.first)) terms.add(l.first);
    }
    if (terms.isEmpty) terms.add(pq.searchTerms.isNotEmpty ? pq.searchTerms.first : 'monster');

    final effGenre = pq.genreEn ?? pq.impliedGenreEn;
    final baseFilter = <String, dynamic>{};
    final code = jwGenreCodes[effGenre] ?? '';
    if (effGenre != null && code.isNotEmpty) baseFilter['genres'] = [code];
    if (pq.yearFrom != null) {
      baseFilter['releaseYear'] = {
        'min': pq.yearFrom,
        'max': pq.yearTo ?? (pq.yearFrom! + 9),
      };
    }

    final movies = <Movie>[];
    final seenKeys = <String>{};
    for (var i = 0; i < terms.take(3).length; i++) {
      final tf = {'searchQuery': terms[i], ...baseFilter};
      List<dynamic> nodes;
      try {
        nodes = await _fetch(tf, baseCountry, n: i == 0 ? 16 : 12, record: i == 0);
      } catch (_) {
        if (i == 0) rethrow;
        continue;
      }
      var added = 0;
      for (final node in nodes) {
        final m = _toMovie(node, pq);
        final keys = m.strongKeys.toSet();
        if (keys.intersection(seenKeys).isNotEmpty) continue;
        seenKeys.addAll(keys);
        movies.add(m);
        added++;
        if (i > 0 && added >= 8) break;
      }
    }

    if (pq.yearFrom != null) {
      movies.removeWhere((m) =>
          m.year != null &&
          !(pq.yearFrom! <= m.year! && m.year! <= (pq.yearTo ?? pq.yearFrom! + 9)));
    }
    if (multiRegion && movies.isNotEmpty) {
      await _mergeExtraRegions(movies, terms.first, baseFilter);
    }
    return movies;
  }

  /// 标题 → 完整条目（供语义通道回链解析；失败不影响健康度）
  Future<List<Movie>> resolveTitle(String title, String country, ParsedQuery? pq) async {
    try {
      final nodes = await _fetch({'searchQuery': title}, country == 'ALL' ? 'US' : country,
          n: 3, record: false);
      return [for (final n in nodes) _toMovie(n, pq)];
    } catch (_) {
      return [];
    }
  }

  Future<void> _mergeExtraRegions(
      List<Movie> movies, String mainTerm, Map<String, dynamic> baseFilter) async {
    final index = <String, Movie>{};
    for (final m in movies) {
      for (final k in m.strongKeys) {
        index.putIfAbsent(k, () => m);
      }
      index.putIfAbsent('t|${_normKey(m.title)}|${m.year}', () => m);
    }
    for (final region in ['GB', 'JP']) {
      List<dynamic> nodes;
      try {
        nodes = await _fetch({'searchQuery': mainTerm, ...baseFilter}, region,
            n: 12, record: false);
      } catch (_) {
        continue;
      }
      for (final node in nodes) {
        final tmp = _toMovie(node, null);
        Movie? target;
        for (final k in tmp.strongKeys) {
          if (index.containsKey(k)) {
            target = index[k];
            break;
          }
        }
        target ??= index['t|${_normKey(tmp.title)}|${tmp.year}'];
        if (target == null) continue;
        final seen = target.offers.map((o) => o.dedupKey).toSet();
        for (final o in tmp.offers) {
          if (!seen.contains(o.dedupKey)) {
            target.offers.add(Offer('${o.platform}·$region', o.kind, o.url, o.price));
            seen.add(o.dedupKey);
          }
        }
        _sortOffers(target.offers);
      }
    }
  }

  Movie _toMovie(dynamic node, ParsedQuery? pq) {
    final n = node as Map;
    final c = (n['content'] ?? {}) as Map;
    final scoring = (c['scoring'] ?? {}) as Map;
    final ext = (c['externalIds'] ?? {}) as Map;
    final genres = [
      for (final g in ((c['genres'] ?? []) as List))
        ((g as Map)['translation'] ?? g['shortName'] ?? '').toString()
    ]..where((s) => s.isNotEmpty).toList();

    final offers = <Offer>[];
    final seen = <String>{};
    for (final o in ((n['offers'] ?? []) as List)) {
      final om = o as Map;
      final pkg = (om['package'] ?? {}) as Map;
      final pkgName = (pkg['clearName'] ?? pkg['shortName'] ?? pkg['technicalName'] ?? '?').toString();
      final kind = _monetizationKind[om['monetizationType']?.toString()] ?? 'other';
      final url = om['standardWebURL']?.toString() ?? '';
      var price = '';
      if (om['retailPrice'] != null) {
        price = '${om['retailPrice']} ${om['currency'] ?? ''}'.trim();
      }
      final off = Offer(pkgName, kind, url, price);
      if (url.isEmpty || seen.contains(off.dedupKey)) continue;
      seen.add(off.dedupKey);
      offers.add(off);
    }
    _sortOffers(offers);

    final fullPath = c['fullPath']?.toString() ?? '';
    var poster = '';
    final posterUrl = c['posterUrl']?.toString();
    if (posterUrl != null && posterUrl.isNotEmpty) {
      poster = '$_imgPrefix$posterUrl'.replaceAll('{profile}', 's332');
    }

    final year = _extYear(c, fullPath);
    final ratings = <String, dynamic>{};
    if (scoring['imdbScore'] != null) ratings['imdb'] = scoring['imdbScore'];

    return Movie(
      title: t2s(c['title']?.toString() ?? ''),
      originalTitle: c['originalTitle']?.toString() ?? '',
      year: year,
      poster: poster,
      overview: (c['shortDescription']?.toString() ?? ''),
      imdbId: ext['imdbId']?.toString() ?? '',
      tmdbId: ext['tmdbId']?.toString() ?? '',
      ratings: ratings,
      imdbVotes: int.tryParse(scoring['imdbVotes']?.toString() ?? '0') ?? 0,
      genres: genres.where((g) => g.isNotEmpty).toList(),
      creatures: pq != null ? List.from(pq.creatures) : [],
      sources: ['justwatch'],
      links: fullPath.isNotEmpty ? {'JustWatch': 'https://www.justwatch.com$fullPath'} : {},
      offers: offers,
    );
  }
}

void _sortOffers(List<Offer> offers) {
  const order = {'free': 0, 'sub': 1, 'rent': 2, 'buy': 3, 'other': 4};
  offers.sort((a, b) {
    final ka = order[a.kind] ?? 9, kb = order[b.kind] ?? 9;
    return ka != kb ? ka.compareTo(kb) : a.platform.compareTo(b.platform);
  });
}

int? _extYear(Map c, String fullPath) {
  final y = c['originalReleaseYear'];
  if (y is int && y > 1880) return y;
  final m = RegExp(r'-(\d{4})$').firstMatch(fullPath.trimRight());
  return m != null ? int.parse(m.group(1)!) : null;
}

String _normKey(String t) => t.toLowerCase().replaceAll(RegExp(r'[\W_]+'), '');

File: backend/app/adapters/justwatch.py
"""JustWatch 适配器（爬虫线 · 核心）。

直接调用其网页端 GraphQL 接口，一次查询同时拿到：
搜索结果 + 海报 + IMDb 评分 + TMDB/IMDb ID + 类型 + 各平台观看地址（含深度链接）。

多语言：searchQuery 原生支持中/日/韩检索词（ゾンビ→活死人之夜），因此对
本体库的 en/ja/ko 主词各发一次查询，本地合并去重，提升母语片名召回率。

接口逆向说明（2026-09 验证有效）：
- 端点禁用 introspection，字段通过 GraphQL 校验错误提示逐个探测；
- 未使用的变量会直接返回 422，查询里的每个变量都必须出现；
- translation(language: "en") 是字符串标量，字面量必须带引号小写；
- searchQuery 多词拼接会稀释结果（模糊全文匹配），只传主词，类型用 genres 短码过滤；
- genre 短码：hrr=恐怖 scf=科幻 trl=惊悚/悬疑 fnt=奇幻 cmy=喜剧 act=动作/冒险 ani=动画 rly=爱情 drm=剧情；
- releaseYear 是 IntFilter，字段名 min/max（服务端年代过滤）；
- offers 的 platform 枚举用 IOS 可返回全部平台报价；
- posterUrl 返回相对路径，{profile} 占位符需替换为尺寸（s332）并加 images 前缀。
"""
from __future__ import annotations

import asyncio
import re
import time

from opencc import OpenCC

from ..models import Movie, Offer
from ..taxonomy import ParsedQuery
from .base import AdapterError, SearchAdapter

try:
    _t2s = OpenCC("t2s")
    _to_simplified = _t2s.convert
except Exception:  # opencc 不可用时保留原文
    _to_simplified = lambda s: s

GQL_URL = "https://apis.justwatch.com/graphql"
IMG_PREFIX = "https://images.justwatch.com"

HEADERS = {
    "Content-Type": "application/json",
    "Origin": "https://www.justwatch.com",
    "Referer": "https://www.justwatch.com/",
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36"
    ),
}

QUERY = """query($f: TitleFilter, $c: Country!, $l: Language!, $p: Platform!, $n: Int!) {
  popularTitles(filter: $f, country: $c, language: $l, first: $n) {
    edges { node {
      id
      content(country: $c, language: $l) {
        title originalTitle fullPath shortDescription
        originalReleaseYear
        scoring { imdbScore imdbVotes }
        externalIds { imdbId tmdbId }
        genres { shortName translation(language: "en") }
        posterUrl(format: JPG)
      }
      offers(country: $c, platform: $p) {
        monetizationType standardWebURL presentationType
        package { shortName technicalName clearName }
        retailPrice(language: $l) currency
      }
    } }
  }
}"""

MONETIZATION_KIND = {
    "FREE": "free", "ADS": "free",       # 免费广告支撑平台（Tubi/Pluto 等）——没会员也能看
    "FLATRATE": "sub",                    # 订阅制（Netflix 等）
    "RENT": "rent", "BUY": "buy",
    "CINEMA": "other", "DOWNLOAD": "buy", "DVD": "other",
}

JW_GENRE_CODES = {
    "horror": "hrr", "thriller": "trl", "sci-fi": "scf", "fantasy": "fnt",
    "comedy": "cmy", "action": "act", "animation": "ani", "adventure": "act",
    "mystery": "trl", "disaster": None,
}


def _norm_key(t: str) -> str:
    import re as _re
    return _re.sub(r"[\W_]+", "", (t or "").lower())


def _year_from_path(path: str) -> int | None:
    m = re.search(r"-(\d{4})$", (path or "").rstrip("/"))
    return int(m.group(1)) if m else None


def ext_year(content: dict, full_path: str) -> int | None:
    y = content.get("originalReleaseYear")
    if isinstance(y, int) and y > 1880:
        return y
    return _year_from_path(full_path)


class JustWatchAdapter(SearchAdapter):
    name = "justwatch"
    kind = "scrape"
    host = "apis.justwatch.com"

    async def _fetch(self, title_filter: dict, country: str, n: int = 24, record: bool = True) -> list[dict]:
        # language=zh：content.title 直接返回中文译名（无译名自动回退原名），
        # originalTitle 保留英文原名，搜索索引与显示语言无关（英文/日文检索词照常用）
        variables = {"f": title_filter, "c": country, "l": "zh", "p": "IOS", "n": n}
        t0 = time.time()
        try:
            r = await self._client.post(
                GQL_URL, json={"query": QUERY, "variables": variables}, headers=HEADERS
            )
            if r.status_code != 200:
                raise AdapterError(f"HTTP {r.status_code}")
            data = r.json()
            if data.get("errors") and not (data.get("data") or {}).get("popularTitles"):
                raise AdapterError(str(data["errors"][0].get("message", ""))[:120])
            edges = ((data.get("data") or {}).get("popularTitles") or {}).get("edges") or []
            if record:
                self.health.record_ok((time.time() - t0) * 1000)
            return [e["node"] for e in edges if e.get("node")]
        except AdapterError as e:
            if record:
                self.health.record_fail(str(e))
            raise
        except Exception as e:
            if record:
                self.health.record_fail(repr(e)[:120])
            raise AdapterError(repr(e)[:120])

    async def search(self, pq: ParsedQuery, country: str) -> list[Movie]:
        # 「所有国家」：美区做发现索引（目录最全），英/日区补查观看渠道后合并
        multi_region = country == "ALL"
        base_country = "US" if multi_region else country
        search_terms = []
        for lang in ("en", "ja", "ko"):
            term = (pq.terms_by_lang.get(lang) or [None])[0]
            if term and term not in search_terms:
                search_terms.append(term)
        if not search_terms:
            search_terms = [pq.search_terms[0] if pq.search_terms else "monster"]

        # 类型：用户显式选择 > 生物隐含类型（消除无类型词时的模糊噪音）
        eff_genre = pq.genre_en or pq.implied_genre_en
        base_filter: dict = {}
        if eff_genre and JW_GENRE_CODES.get(eff_genre):
            base_filter["genres"] = [JW_GENRE_CODES[eff_genre]]
        if pq.year_from:
            base_filter["releaseYear"] = {"min": pq.year_from, "max": pq.year_to or pq.year_from + 9}

        # 主语言（en）全量；非主语言限量补充（去重后每语言最多 8 条新片）
        movies: list[Movie] = []
        seen_keys: set = set()

        def _keys(m: Movie):
            ks = {k for k in m.strong_keys()}
            return ks

        for i, term in enumerate(search_terms[:3]):
            tf = {"searchQuery": term, **base_filter}
            try:
                nodes = await self._fetch(tf, base_country, n=16 if i == 0 else 12,
                                          record=(i == 0))
            except AdapterError:
                if i == 0:
                    raise
                continue
            added = 0
            for node in nodes:
                m = self._to_movie(node, pq)
                ks = _keys(m)
                if ks & seen_keys:
                    continue
                seen_keys |= ks
                movies.append(m)
                added += 1
                if i > 0 and added >= 8:
                    break

        if pq.year_from:  # 双保险：本地再过滤一次（无年份的保留）
            movies = [
                m for m in movies
                if m.year is None or pq.year_from <= m.year <= (pq.year_to or pq.year_from + 9)
            ]
        if multi_region and movies:
            await self._merge_extra_regions(movies, search_terms[0], base_filter)
        return movies

    async def _merge_extra_regions(self, movies: list[Movie], main_term: str, base_filter: dict) -> None:
        """「所有国家」模式：英/日区补查观看渠道，合并进已命中的影片（平台名带地区标记）。"""
        index: dict[str, Movie] = {}
        for m in movies:
            for k in m.strong_keys():
                index.setdefault(k, m)
            index.setdefault(f"t|{_norm_key(m.title)}|{m.year}", m)

        for region in ("GB", "JP"):
            try:
                nodes = await self._fetch({"searchQuery": main_term, **base_filter},
                                          region, n=12, record=False)
            except Exception:
                continue
            for node in nodes:
                tmp = self._to_movie(node, None)
                target = None
                for k in tmp.strong_keys():
                    if k in index:
                        target = index[k]
                        break
                if target is None:
                    target = index.get(f"t|{_norm_key(tmp.title)}|{tmp.year}")
                if target is None:
                    continue
                seen = {o.dedup_key() for o in target.offers}
                for o in tmp.offers:
                    if o.dedup_key() not in seen:
                        o.platform = f"{o.platform}·{region}"
                        target.offers.append(o)
                        seen.add(o.dedup_key())
                target.offers.sort(key=lambda o: (
                    {"free": 0, "sub": 1, "rent": 2, "buy": 3, "other": 4}.get(o.kind, 9), o.platform))

    async def resolve_title(self, title: str, country: str, pq: ParsedQuery | None = None) -> list[Movie]:
        """标题 → 完整条目（含观看地址）。供 Yandex 等发现型通道回链解析用，
        失败不影响本源健康度。"""
        try:
            nodes = await self._fetch({"searchQuery": title}, country, n=3, record=False)
        except Exception:
            return []
        return [self._to_movie(n, pq) for n in nodes]

    def _to_movie(self, node: dict, pq: ParsedQuery | None) -> Movie:
        c = node.get("content") or {}
        scoring = c.get("scoring") or {}
        ext = c.get("externalIds") or {}
        genres = []
        for g in c.get("genres") or []:
            genres.append(g.get("translation") or g.get("shortName") or "")

        offers, seen = [], set()
        for o in node.get("offers") or []:
            pkg = o.get("package") or {}
            pkg_name = pkg.get("clearName") or pkg.get("shortName") or pkg.get("technicalName") or "?"
            kind = MONETIZATION_KIND.get(o.get("monetizationType") or "", "other")
            url = o.get("standardWebURL") or ""
            price = ""
            if o.get("retailPrice") is not None:
                price = f"{o.get('retailPrice')} {o.get('currency') or ''}".strip()
            off = Offer(platform=pkg_name, kind=kind, url=url, price=price)
            if not url or off.dedup_key() in seen:
                continue
            seen.add(off.dedup_key())
            offers.append(off)
        offers.sort(key=lambda o: (
            {"free": 0, "sub": 1, "rent": 2, "buy": 3, "other": 4}.get(o.kind, 9), o.platform))

        full_path = c.get("fullPath") or ""
        poster = ""
        if c.get("posterUrl"):
            poster = f"{IMG_PREFIX}{c['posterUrl']}".replace("{profile}", "s332")

        return Movie(
            title=_to_simplified(c.get("title") or ""),
            original_title=c.get("originalTitle") or "",
            year=ext_year(c, full_path),
            poster=poster,
            overview=(c.get("shortDescription") or "")[:300],
            imdb_id=ext.get("imdbId") or "",
            tmdb_id=str(ext.get("tmdbId") or ""),
            ratings={"imdb": scoring.get("imdbScore")},
            imdb_votes=int(scoring.get("imdbVotes") or 0),
            genres=[g for g in genres if g],
            creatures=list(pq.creatures) if pq else [],
            sources=["justwatch"],
            links={"JustWatch": f"https://www.justwatch.com{full_path}"} if full_path else {},
            offers=offers,
        )
